drivers/powervar-hid.c

```diff
@@ -0,0 +1,146 @@
+/* --------------------------------------------------------------- */
+
+/* POWERVAR usage table */
+static usage_lkp_t powervar_usage_lkp[] = {
+	{ "POWERVAR1",	0xff000001 },
+	{  NULL, 0 }
+};
+
+static usage_tables_t powervar_utab[] = {
+	powervar_usage_lkp,
+	hid_usage_lkp,
+	NULL,
+};
+
+/* --------------------------------------------------------------- */
+/* HID2NUT lookup table                                            */
+/* --------------------------------------------------------------- */
+
+static hid_info_t powervar_hid2nut[] = {
+
+  { "ups.powersummary.iproduct", 0, 0, "UPS.PowerSummary.iProduct", NULL, "%.0f", 0, NULL },
+  { "ups.powersummary.iserialnumber", 0, 0, "UPS.PowerSummary.iSerialNumber", NULL, "%.0f", 0, NULL },
+  { "ups.powersummary.imanufacturer", 0, 0, "UPS.PowerSummary.iMnaufacturer", NULL, "%.0f", 0, NULL },
+  /* Battery page */
+  { "battery.type", 0, 0, "UPS.PowerSummary.iDeviceChemistry", NULL, "%s", 0, stringid_conversion },
+  { "battery.ioeminformation", 0, 0, "UPS.PowerSummary.iOEMInformation", NULL, "%s", 0, stringid_conversion },
+  { "battery.capacitymode", 0, 0, "UPS.PowerSummary.CapacityMode", NULL, "%.0f", 0, NULL },
+  { "battery.charge", 0, 0, "UPS.PowerSummary.RemainingCapacity", NULL, "%.0f", 0, NULL },
+  { "battery.capacity", 0, 0, "UPS.PowerSummary.FullChargeCapacity", NULL, "%.0f", 0, NULL },
+  { "battery.designcapacity", 0, 0, "UPS.PowerSummary.DesignCapacity", NULL, "%.0f", 0, NULL },
+  { "battery.runtime", 0, 0, "UPS.PowerSummary.RuntimeToEmpty", NULL, "%.0f", 0, NULL },
+
+/* Special case: ups.status & ups.alarm */
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.ACPresent", NULL, NULL, HU_FLAG_QUICK_POLL, online_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.Discharging", NULL, NULL, HU_FLAG_QUICK_POLL, discharging_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.Charging", NULL, NULL, HU_FLAG_QUICK_POLL, charging_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.ShutdownImminent", NULL, NULL, 0, shutdownimm_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.BelowRemainingCapacityLimit", NULL, NULL, HU_FLAG_QUICK_POLL, lowbatt_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.Overload", NULL, NULL, 0, overload_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.NeedReplacement", NULL, NULL, 0, replacebatt_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.RemainingTimeLimitExpired", NULL, NULL, 0, timelimitexpired_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.InternalFailure", NULL, NULL, 0, yes_no_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.Good", NULL, NULL, 0, yes_no_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.OverTemperature", NULL, NULL, 0, overheat_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.FullyCharged", NULL, NULL, 0, fullycharged_info },
+  { "BOOL", 0, 0, "UPS.PowerSummary.PresentStatus.FullyDischarged", NULL, NULL, 0, depleted_info },
```

Comment:
@andrewmccartney are these all the usage paths that you get when running usbhid-ups -DDD -x explore
see also #807 on settings (battery.charge.low for ex) and commands

Reply:
@aquette Yes, these are the items we are supporting at this time. I don't think we have any commands or settings at this time. Please advise if I'm mis-understanding your comment about settings and commands.

Reply:
At least the shutdown and startup delays usages should be implemented to reach the UPS power off feature.
I'd still like to see the debut output

Reply:
Here is the debug output as requested

 228.233213	[D3] usb_busses=0x559d6bd44e60
 228.233229	[D2] Checking device (4234/0002) (002/075)
 228.234544	[D2] - VendorID: 4234
 228.234566	[D2] - ProductID: 0002
 228.234573	[D2] - Manufacturer: AMETEK-POWERVAR Corporation
 228.234580	[D2] - Product: UPM UPS (v01.51, Nov  3 2018)
 228.234586	[D2] - Serial Number: 5014403R-1740135
 228.234619	[D2] - Bus: 002
 228.234634	[D2] - Device release number: 0002
 228.234644	[D2] Set rep index to 1
 228.234658	[D2] Trying to match device
 228.234675	[D2] Device matches
 228.234691	[D2] failed to claim USB device: could not claim interface 1: Device or resource busy
 228.234760	[D2] detached kernel driver from USB device...
 228.234787	[D3] nut_usb_set_altinterface: skipped usb_set_altinterface(udev, 0)
 228.235082	[D3] Report[get]: (2 bytes) => 02 02
 228.235111	[D2] Path: UPS.PowerSummary.iProduct, Type: Feature, ReportID: 0x02, Offset: 0, Size: 8, Value: 2
 228.235331	[D3] Report[get]: (2 bytes) => 03 04
 228.235359	[D2] Path: UPS.PowerSummary.iSerialNumber, Type: Feature, ReportID: 0x03, Offset: 0, Size: 8, Value: 4
 228.235373	[D3] Report[buf]: (2 bytes) => 02 02
 228.235387	[D2] Path: UPS.PowerSummary.iMnaufacturer, Type: Feature, ReportID: 0x02, Offset: 0, Size: 8, Value: 2
 228.235582	[D3] Report[get]: (2 bytes) => 04 06
 228.235609	[D2] Path: UPS.PowerSummary.iDeviceChemistry, Type: Feature, ReportID: 0x04, Offset: 0, Size: 8, Value: 6
 228.236330	[D3] Report[get]: (2 bytes) => 06 01
 228.236357	[D2] Path: UPS.PowerSummary.iOEMInformation, Type: Feature, ReportID: 0x06, Offset: 0, Size: 8, Value: 1
 228.237057	[D3] Report[get]: (2 bytes) => 07 02
 228.237076	[D2] Path: UPS.PowerSummary.CapacityMode, Type: Feature, ReportID: 0x07, Offset: 0, Size: 8, Value: 2
 228.237307	[D3] Report[get]: (2 bytes) => 0c 64
 228.237326	[D2] Path: UPS.PowerSummary.RemainingCapacity, Type: Feature, ReportID: 0x0c, Offset: 0, Size: 8, Value: 100
 228.237558	[D3] Report[get]: (2 bytes) => 09 64
 228.237576	[D2] Path: UPS.PowerSummary.FullChargeCapacity, Type: Feature, ReportID: 0x09, Offset: 0, Size: 8, Value: 100
 228.238248	[D3] Report[get]: (2 bytes) => 08 64
 228.238272	[D2] Path: UPS.PowerSummary.DesignCapacity, Type: Feature, ReportID: 0x08, Offset: 0, Size: 8, Value: 100
 228.238418	[D3] Report[get]: (4 bytes) => 0e 40 7e 05
 228.238434	[D2] Path: UPS.PowerSummary.RuntimeToEmpty, Type: Feature, ReportID: 0x0e, Offset: 0, Size: 24, Value: 360000
 228.238686	[D3] Report[get]: (3 bytes) => 0a 04 09
 228.238712	[D2] Path: UPS.PowerSummary.PresentStatus.ACPresent, Type: Feature, ReportID: 0x0a, Offset: 8, Size: 1, Value: 1
 228.238729	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238744	[D2] Path: UPS.PowerSummary.PresentStatus.Discharging, Type: Feature, ReportID: 0x0a, Offset: 13, Size: 1, Value: 0
 228.238758	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238769	[D2] Path: UPS.PowerSummary.PresentStatus.Charging, Type: Feature, ReportID: 0x0a, Offset: 12, Size: 1, Value: 0
 228.238784	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238797	[D2] Path: UPS.PowerSummary.PresentStatus.ShutdownImminent, Type: Feature, ReportID: 0x0a, Offset: 1, Size: 1, Value: 0
 228.238810	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238822	[D2] Path: UPS.PowerSummary.PresentStatus.BelowRemainingCapacityLimit, Type: Feature, ReportID: 0x0a, Offset: 10, Size: 1, Value: 0
 228.238837	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238848	[D2] Path: UPS.PowerSummary.PresentStatus.Overload, Type: Feature, ReportID: 0x0a, Offset: 3, Size: 1, Value: 0
 228.238860	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238874	[D2] Path: UPS.PowerSummary.PresentStatus.NeedReplacement, Type: Feature, ReportID: 0x0a, Offset: 15, Size: 1, Value: 0
 228.238897	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238910	[D2] Path: UPS.PowerSummary.PresentStatus.InternalFailure, Type: Feature, ReportID: 0x0a, Offset: 0, Size: 1, Value: 0
 228.238924	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238936	[D2] Path: UPS.PowerSummary.PresentStatus.Good, Type: Feature, ReportID: 0x0a, Offset: 2, Size: 1, Value: 1
 228.238950	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238964	[D2] Path: UPS.PowerSummary.PresentStatus.OverTemperature, Type: Feature, ReportID: 0x0a, Offset: 4, Size: 1, Value: 0
 228.238977	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.238989	[D2] Path: UPS.PowerSummary.PresentStatus.FullyCharged, Type: Feature, ReportID: 0x0a, Offset: 11, Size: 1, Value: 1
 228.239001	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239013	[D2] Path: UPS.PowerSummary.PresentStatus.FullyDischarged, Type: Feature, ReportID: 0x0a, Offset: 14, Size: 1, Value: 0
 228.239025	[D1] Not using interrupt pipe...
 228.239037	[D1] Quick update...
 228.239049	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239060	[D2] Path: UPS.PowerSummary.PresentStatus.ACPresent, Type: Feature, ReportID: 0x0a, Offset: 8, Size: 1, Value: 1
 228.239072	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239082	[D2] Path: UPS.PowerSummary.PresentStatus.Discharging, Type: Feature, ReportID: 0x0a, Offset: 13, Size: 1, Value: 0
 228.239092	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239104	[D2] Path: UPS.PowerSummary.PresentStatus.Charging, Type: Feature, ReportID: 0x0a, Offset: 12, Size: 1, Value: 0
 228.239118	[D3] Report[buf]: (3 bytes) => 0a 04 09
 228.239130	[D2] Path: UPS.PowerSummary.PresentStatus.BelowRemainingCapacityLimit, Type: Feature, ReportID: 0x0a, Offset: 10, Size: 1, Value: 0
 230.234654	[D1] upsdrv_updateinfo...

Reply:
Thx.
Indeed, none of the standard delay to power off the unit. I may have missed something, but are you manufacturing it? If so, how do you achieve the power off in the end?
Otherwise looks good and can be merged

Reply:
With regards to the shutdown and startup delays. At this time we have not implemented them in our UPS for HID-battery, so I am assuming that it does not make sense to include them in our NUT driver. Is this correct?

Reply:
Looks like our comments were made at the same time ... we do have various shutdown controls in the UPS (timed, RS232 shutdown pin, load based, and of course low battery voltage) but none of them are included in our HID-battery implementation at this time. So if you shutdown the system that is connected to the UPS it will continue to idle on battery until either AC returns or the battery is exhausted, unless you have configured one of the shutdown mechanisms I listed above outside of the HID-battery interface. For example, configure the UPS to shutdown if on battery and the load falls below a minimum level or a timed shutdown from when it transfers to battery. Hopefully this makes sense.

Reply:
Indeed, so good to me.
And whenever you add FW features, don't forget to update your nut subdriver